gdal/port/userfaultfd.cpp

```diff
@@ -0,0 +1,136 @@
+ * Copyright (c) 2018, James McClain <james.mcclain@gmail.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include <cstdlib>
+#include <cstring>
+
+#include <fcntl.h>
+#include <pthread.h>
+#include <unistd.h>
+
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <linux/userfaultfd.h>
+
+#include "cpl_vsi.h"
+#include "userfaultfd.h"
+
+
+struct args {
+  int uffd;
+  char filename[0x100];
+  uint64_t page_size;
+  void * memory;
+  off_t memory_size;
+};
+
+void * fault_handler(void * ptr)
+{
+  struct args * args = static_cast<struct args *>(ptr);
+  struct uffd_msg msg;
+  struct uffdio_copy uffdio_copy;
+  void * page = mmap(NULL, args->page_size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0); // XXX
+
+  // Open file for reading
+  VSILFILE * file = VSIFOpenL(args->filename, "rb");
+
+  fprintf(stderr, "%s:%d\n", __FILE__, __LINE__);
+
+  // Loop indefinitely
+  while(1) {
+    if (read(args->uffd, &msg, sizeof(msg)) < 1) break;
+    fprintf(stderr, "%s:%d\n", __FILE__, __LINE__);
+    uintptr_t fault_addr = msg.arg.pagefault.address & ~(args->page_size-1);
+    uint64_t offset = (uint64_t)fault_addr - (uint64_t)args->memory;
+
+    // Copy data into page
+    VSIFSeekL(file, offset, SEEK_SET);
+    VSIFReadL(page, args->page_size, 1, file);
+
+    uffdio_copy.src = (uintptr_t)(page); // XXX
+    uffdio_copy.dst = fault_addr;
+    uffdio_copy.len = (uintptr_t)(args->page_size); // XXX
+    uffdio_copy.mode = 0;
+    uffdio_copy.copy = 0;
+    if (ioctl(args->uffd, UFFDIO_COPY, &uffdio_copy) == -1) break;
+  }
+
+  // Return resources
+  VSIFCloseL(file);
+  close(args->uffd);
+  munmap(page, args->page_size);
+
+  return nullptr;
+}
+
+void hookup(const char * pszFilename, void ** ppMemory, off_t * pSize)
```

Comment:
a better name would be needed: CPLCreateUserFaultMapping() ?
an error status should be returned

Reply:
Okay, will change.